wordpress-mobile/WordPress-FluxC-Android
# Add initial support for experimentation platform

This PR adds an initial support for the Experimentation Platform API (internal ref: pbmo2S-yK-p2).

## Usage

The main class added by this PR is the `ExperimentStore`, to which a client implementation would interact in the following way:

First, it would fetch the latest experiment assignments for an user by calling `experimentStore::fetchAssignments`:

```kotlin
val payload = FetchAssignmentsPayload(Platform.WORDPRESS_ANDROID)
experimentStore.fetchAssignments(payload)
```

And then handle the resulting `OnAssignmentsFetched`:

```kotlin
fun onAssignmentsFetched(result: OnAssignmentsFetched) {
    if (!result.isError) {
        val assignments = event.assignments
        // Handle assignments
    } else {
        // Handle error
    }
}
```

Once the assignments are fetched, they are automatically cached and can be retrieved like so:

```kotlin
val assignments = experimentStore.getCachedAssignments()
```

With an `Assignments` object in hand, it's possible to retrieve a `Variation` for a specific experiment and then handle it accordingly:

```kotlin
when (val variation = assignments.getVariationForExperiment("my_experiment")) {
    is Control -> handleControl() // Show control option
    is Treatment -> handleTreatment() // Show treatment option
    is Unknown -> handleUnknown() // Experiment doesn't exist or wasn't retrieved
    is Other -> when (variation.name) { // Any other value
        "other_value_one" -> handleValueOne()
        "other_value_two" -> handleValueTwo()
        else -> handleUnknown()
    }
}
```

It's also possible to check if we should fetch newer assignments:

```kotlin
if (assignments.isStale()) {
    // TTL has expired, we should fetch newer assignments for this user
}
```

## Next steps

Theoretically, this implementation already gives us the ability to use the Experimentation Platform API in our apps, although we still need to wait for the apps to be officially supported by the platform (more information about this in the internal reference linked above).

From a practical perspective however, we should probably still add a middle abstraction between the clients and this implementation to make it a little bit easier to do stuff like generating and storing anonymous IDs and automatically fetching new values when the cached results are stale, etc.

The way I see it, the most obvious candidates to receive such implementation would be either [WordPress-Analytics-Android](https://github.com/wordpress-mobile/WordPress-Analytics-Android), a brand new library, or even the apps themselves if we don't care much about sharing that logic. 

This is totally open for discussion.

## To test

You can either run the included unit tests or use the following steps on the example app:

1. On the initial screen, tap **Experiments**.
1. On the next screen, select a platform to query (keep in mind the only platforms officially supported at the moment are `calypso` and `wpcom`).
1. Optionally, you can add or generate an Anonymous ID to be included in the request as well.
1. Tap **Fetch Assignments**.
1. Check the results printed by the in-app console.
1. Tap **Get Cached Assignments**.
1. Verify that the returned results match the ones fetched before.

Feel free to try out different combinations, both signed in and signed out.
